File: problems/python/32. Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        l, r, res = 0, 0, 0
        
        for c in s:
            if c == '(':
                l += 1
            else:
                r += 1
            if l == r:
                res = max(res, 2 * r)
            elif r > l:
                l, r = 0, 0
        
        l, r = 0, 0
        for i in range(len(s) - 1, -1, -1):
            c = s[i]
            if c == '(':
                l += 1
            else:
                r += 1
            if l == r:
                res = max(res, 2 * l)
            elif l > r:
                l, r = 0, 0
                
        return res

File: problems/C++/32. Longest Valid Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int l = 0, r = 0, res = 0;

        for (char c : s) {
            if (c == '(') {
                l++;
            } else {
                r++;
            }
            if (l == r) {
                res = max(res, 2 * r);
            } else if (r > l) {
                l = 0;
                r = 0;
            }
        }

        l = 0;
        r = 0;
        for (int i = s.length() - 1; i >= 0; --i) {
            char c = s[i];
            if (c == '(') {
                l++;
            } else {
                r++;
            }
            if (l == r) {
                res = max(res, 2 * l);
            } else if (l > r) {
                l = 0;
                r = 0;
            }
        }
        return res;
    }
};
